### ISSUE TITLE
NewConnectionError 

### ISSUE THREAD

**Issue by JaswantD** [jepcastelein/marketo-rest-python#50]
Hi, 

I had written a script earlier to fetch programs and it's members data using your library and the script was running fine till last Friday. But now it is giving me the following error:

HTTPSConnectionPool(host='***-***-***.mktorest.com', port=443): Max retries exceeded with url: /rest/v1/leads/programs/1340.json?access_token=ccbb3359-4a6e-4f00-ae8e-f831bb859c75%3Asj&_method=GET&nextPageToken=SW3PTMBVFCNHSHJGZ7LQH3ZWNXMGVWHK2OFL43GOERUWKVM7LLML6QDMXPPMMWE5WDMS2VAF7532E%3D%3D%3D (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x000001F1288A54E0>: Failed to establish a new connection: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted',))

Could you help me on this??

**Comment by jepcastelein** [comment]
I have not seen this error before. Based on a quick Google search, is it
possible that you have the script running twice on the same server?
https://stackoverflow.com/questions/45767905/errno-10048-only-one-usage-of-each-socket-address-protocol-network-address-p?rq=1

**Comment by JaswantD** [comment]
This is strange for us too that the script had been running from the last six to seven months without any error, but we suddenly started getting this error from last Friday.

**Comment by JaswantD** [comment]
Hi, Still getting the same issues in every script where I am using rest API call. Could you please look into this and help to solve this issue.

**Comment by jepcastelein** [comment]
The library hasn't changed in 5 months, so I assume there is an environment
issue. Were you able to successfully run it on another machine?

**Comment by JaswantD** [comment]
No, I am getting the same issue in every machine where I am using this library from last week, Even on clients' servers. Yesterday we tried to pull data of 200 programs with its members using PHP and it ran successfully. So I doubt why we are getting this using python. 
I tried by updating my python version from 3.6 to 3.7.4. but it didn't help.

**Comment by jepcastelein** [comment]
what is the marketorestpython version you are using? What is your Python
code?

**Comment by JaswantD** [comment]
I am using marketorestpython version - 0.5. You can check my code from the below link-

https://github.com/JaswantD/MarketoRest/blob/master/Programs_Marketo.py

**Comment by jepcastelein** [comment]
I would try to reproduce/isolate the issue by first identifying which
program_id the error occurs for, then writing a super basic script that
only does this:

members = mc.execute(method='get_multiple_leads_by_program_id',
programId=id,
                                               fields=['createdAt',
'email', 'firstName', 'id', 'lastName',
                                                       'leadStatus',
'sfdcAccountId', 'sfdcContactId', 'sfdcId',
                                                       'sfdcLeadId',
'updatedAt', 'mktoName', 'title', 'company',
                                                       'phone',
                                                       'leadSource',
'sfdcType', 'True_Market_Name__c',
                                                       'Market_Region__c',
'Lead_Source_Detail__c',

 'Account_Owner_Email__c'], batchSize=None)

If that still results in an error, it should be much easier to debug.

**Comment by JaswantD** [comment]
Hi,

is this possible that our library may be making multiple connections? As per last line of error :

"HTTPSConnectionPool(host='021-ncn-364.mktorest.com', port=443): Max retries exceeded with url: /identity/oauth/token?grant_type=client_credentials&client_id=59826789-95f4-4de5-b1be-f6a130361b82&client_secret=nB5AGURk8rTKfIGlfT7Q4lCn1IDfC7Ay (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x000001E4378BC6A0>: **Failed to establish a new connection: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted**',))

**Comment by jepcastelein** [comment]
Not that I know of, but feel free to review the code. I would also consider
using 'get_multiple_leads_by_program_id_yield, which returns in batches of
300 records, rather than looping and returning all program members at once.

**Comment by JaswantD** [comment]
Hi,  I got following response from Marketo on this issue:
"Unfortunately, since the python library you are using is not created or maintained by Marketo we cannot guarantee the results or help troubleshoot errors encountered when using it, However I made some API calls myself directly using Postman and I was able to confirm the results are being returned as expected when trying to fetch data for program 1340 / 3260.

According to most of the resources I was able to find, the issue seems to be resolved after closing the socket implementing serve_forever()."

**Comment by data-alchemist** [comment]
@JaswantD , I am facing similar error. The code was working all right till a month ago. But now it started throwing `Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known`

Were you able to solve it?
